docs: update locators doc to use new APIs (#18352)
microsoft/playwright: docs/src/locators.md

```

```python sync
page.get_by_role("button").count()
```

```java
page.getByRole("button").count();
```

```csharp
await page.GetByRole("button").CountAsync();
```

You can explicitly opt-out from strictness check by telling Playwright which element to use when multiple element match, through [`method: Locator.first`], [`method: Locator.last`], and [`method: Locator.nth`]. These methods are **not recommended** because when your page changes, Playwright may click on an element you did not intend. Instead, follow best practices below to create a locator that uniquely identifies the target element.

## Locating elements

Playwright comes with multiple built-in ways to create a locator. To make tests resilient, we recommend prioritizing user-facing attributes and explicit contracts, and provide dedicated methods for them, such as [`method: Page.getByText`]. It is often convenient to use the [code generator](./codegen.md) to generate a locator, and then edit it as you'd like.

If you absolutely must use CSS or XPath locators, you can use [`method: Page.locator`] to create a locator that takes a [selector](./selectors.md) describing how to find an element in the page.

> Note that all methods that create a locator, such as [`method: Page.getByLabel`], are also available on the [Locator] and [FrameLocator] classes, so you can chain them and iteratively narrow down your locator.

Reviewer: would love an example of this as I am not quite sure what it means. isn't getByLabel a locator so ... framelocator ? API docs has a nice example we could possibly use here:
```js
const locator = page.frameLocator('#my-frame').getByText('Submit');
await locator.click();
```

> [`method: Page.locator`] is the most generic way to create a locator that takes a [selector](./selectors.md) describing how to find an element in the page, such as a CSS or XPath selector. That said, we highly recommend to use dedicated methods when you can, and rely on [`method: Page.locator`] as a fallback when more resilient locators are not possible. When in doubt, consider using [code generator](./codegen.md) to generate a locator for you.

Reviewer 2: "is the most generic way" suggests that this is a preferred option.

```suggestion
If you absolutely must use CSS or XPath locators, you can use [`method: Page.locator`] to create a locator that takes a [selector](./selectors.md) describing how to find an element in the page.
```

page.get_by_role("checkbox", checked=True, name="Check me"]).check()
```

```java
page.getByRole("button", new Page.GetByRoleOptions().setName(Pattern.compile("submit", Pattern.CASE_INSENSITIVE))).click();

page.getByRole("checkbox", new Page.GetByRoleOptions().setChecked(true).setName("Check me"))).check();
```

```csharp
await page.GetByRole("button", new() { Name = new Regex("submit", RegexOptions.IgnoreCase) }).ClickAsync();

await page.GetByRole("checkbox", new() { Checked = true, Name = "Check me" }).CheckAsync();
```

Role locators follow W3C specificaitons for [ARIA role](https://www.w3.org/TR/wai-aria-1.2/#roles), [ARIA attributes](https://www.w3.org/TR/wai-aria-1.2/#aria-attributes) and [accessible name](https://w3c.github.io/accname/#dfn-accessible-name).

Note that role locators **do not replace** accessibility audits and conformance tests, but rather give early feedback about the ARIA guidelines.

### Define explicit contract and use [`method: Page.getByTestId`]

User-facing attributes like text or accessible name can change over time. In this case it is convenient to define explicit test ids.

```html

> <button data-test-id="directions">Itinéraire</button>

Reviewer: our api docs say: By default, the data-testid attribute is used as a test id. so should we be using data-testid or data-test-id` or are both acceptable?

Author: Ouch, my bad!

```

```js
await page.getByTestId('directions').click();
```

```java
page.getByTestId("directions").click();
```

```python async
await page.get_by_test_id('directions').click()
```

```python sync
page.get_by_test_id('directions').click()
```

```csharp
await page.GetByTestId("directions").ClickAsync();
```

> By default, [`method: Page.getByTestId`] will locate elements baed on the `data-test-id` attribute, but you can configure it in your test config or calling [`method: Selectors.setTestIdAttribute`].

Reviewer: same as above data-testid or data-test-id

### Locate by label text with [`method: Page.getByLabel`]

Most form controls usually have dedicated labels that could be conveniently used to interact with the form. In this case, you can locate the control by its associated label.

For example, consider the following DOM structure.

```html
<label for="password">Password:</label><input type="password">
```

You can fill the input after locating it by the label text:

```js
await page.getByLabel('Password').fill('secret');
```

```java
page.getByLabel("Password").fill("secret");
```

```python async
await page.get_by_label("Password").fill("secret")
```

```python sync
page.get_by_label("Password").fill("secret")
```

```csharp
await page.GetByLabel("Password").FillAsync("secret");
```

### Locate in a subtree

You can chain methods that create a locator, like [`method: Page.getByText`] or [`method: Locator.getByRole`], to narrow down the search to a particular part of the page.

For example, consider the following DOM structure:

> 

Reviewer: The example here has data-test-id in the html so if we need to change this to data-testid then keep this in mind.

If you have a list of identical elements, and the only way to distinguish between them is the order, you can choose a specific element from a list with [`method: Locator.first`], [`method: Locator.last`] or [`method: Locator.nth`].

For example, to click the third item in the list of products:

```js
await page.getByTestId('product-card').nth(3).click();
```

```java
page.getByTestId("product-card").nth(3).click();
```

```python async
await page.get_by_test_id("product-card").nth(3).click()
```

```python sync
page.get_by_test_id("product-card").nth(3).click()
```

```csharp
await page.GetByTestId("product-card").Nth(3).ClickAsync();
```

However, use these methods with caution. Often times, the page might change, and locator will point to a completely different element from the one you expected. Instead, try to come up with a unique locator that will pass the [strictness criteria](#strictness).

## Locator vs ElementHandle

> :::caution

Reviewer: suggestion to move locator vs elementHandle to handles doc so that it isnt something people read or find unless they look for it (as discussed with Pavel earlier)

Author: Sounds good.
